chore/check-javadoc-regressions.py

```diff
@@ -0,0 +1,272 @@
+    a file it will find the exact match though.
+    """
+    output: List[str] = []
+
+    for line in with_numbers:
+        stripped = strip_line_number(line)
+        if stripped in without_numbers:
+            if without_numbers[stripped] == 1:
+                without_numbers.pop(stripped)
+            else:
+                without_numbers[stripped] -= 1
+            output.append(line)
+
+    for line in without_numbers:
+        output.append(line)
+
+    return output
+
+
+def print_regression_lines(reference: str, other: str) -> None:
+    """
+    Receives two checkstyle outputs, one for the reference and one for the current branch and
+    prints all identified regressions.
+    """
+    lines_reference = filter_relevant_lines(reference.splitlines())
+    lines_other = filter_relevant_lines(other.splitlines())
+
+    new_errors = find_added_violation_lines(
+        Counter(map(strip_line_number, lines_reference)),
+        Counter(map(strip_line_number, lines_other))
+    )
+    with_line_numbers = try_readd_line_numbers(new_errors, lines_other)
+
+    print(warn("You likely added the following new errors. "), end=" ")
+    print("Line numbers might be incorrect if there are multiple violations of the same type in the file.")
+    for error in with_line_numbers:
+        print(error.strip())
+
+
+def print_compare_with_branch(target_branch: str) -> None:
+    """
+    Compares the current branch with a given reference branch and prints a change summary.
+    Exits with an error if the current branch has more violations than the passed target branch.
+    """
+    # Switch to the root so checkstyle will run against the whole project
+    os.chdir(run_command(["git", "rev-parse", "--show-toplevel"]).strip())
+
+    with temporary_path(suffix=".xml") as config_path:
+        write_checkstyle_config(config_path)
+
+        run_command(["git", "checkout", target_branch], stderr=PIPE)
+        reference_output = run_checkstyle(config_path)
+        run_command(["git", "checkout", "-"], stderr=PIPE)
+        other_output = run_checkstyle(config_path)
+
+        reference_violation_count = extract_violation_count(reference_output)
+        other_violation_count = extract_violation_count(other_output)
+
+        if reference_violation_count is None:
+            print(warn("The Checkstyle run for the reference branch did not yield any result."))
+            exit(1)
+        if other_violation_count is None:
+            print(warn("The Checkstyle run for your branch did not yield any result."))
+            exit(1)
+
+        print()
+        print("Current status")
+        print(
+            f"  Violations on {hl('comparison')} branch ({target_branch}): "
+            f"{hl(str(reference_violation_count))}")
+        print(
+            f"  Violations on {hl('your')} branch         {' '*len(target_branch)}: "
+            f"{hl(str(other_violation_count))}"
+        )
+        print()
```

Comment:
Why not just print one large string here? I think that'd make it easier to read (in the source code, that is).

```python
        print(f"""
Current status
  Violations on {hl('comparison')} branch ({target_branch}): {hl(str(reference_violation_count))}
  Violations on {hl('your')} branch         {' '*len(target_branch)}: {hl(str(other_violation_count))}
""")
```

I'd also suggest making these large blobs of text prints separate functions, so it the source reads e.g. `print_status(target_branch, reference_violation_count, other_violation_count)`

Reply:
Maybe that is nicer, I just don't really like that pythons text blocks consider *all* the whitespace to the left, forcing you to break the indentation of your program. This wouldn't be as bad in Java, but as indentation is semantically important in Python it feels a bit off to have a string hanging out at the start of the line.

I've changed it though.

Reply:
> I just don't really like that pythons text blocks consider all the whitespace to the left

That is a bit annoying, but working around it is pretty easy if one is prone to be bothered by such things.

```python
def dedent(s: str, start_char: str = "|") -> str:
    return "\n".join([substring(line, start_char) for line in s.split("\n") ])

def substring(s: str, from_char: str) -> str:
    start_idx = s.find(from_char) + 1
    return s[start_idx : ]

s = """Hello there
    |This is just a little test
    |of dedenting a string
"""

print(dedent(s))
```

That outputs 

```
Hello there
This is just a little test
of dedenting a string
```

But yeah, a bit overkill for a support script like this.

Reply:
Thank you! :) Yea, it is probably fine here. The strings only appear in one place and it isn't too hard to follow, so I will probably keep your initial suggestion and leave them as-is.

Sometimes you just get a bit spoiled by features in other languages...